x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticationService.java

```diff
@@ -690,6 +694,45 @@ void writeAuthToContext(Authentication authentication) {
         private void authenticateToken(AuthenticationToken token) {
             this.consumeToken(token);
         }
+
+        private User maybeMergeAnonymousRolesForUser(User user) {
+            if (SystemUser.is(user) || XPackUser.is(user) || XPackSecurityUser.is(user) || AsyncSearchUser.is(user)) {
```

Comment:
I'm also a bit worried about duplicating the check here. Should we make https://github.com/elastic/elasticsearch/blob/9467dbf1208970b7e8e60d3650b53dc0e6151514/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java#L418 public, update it to take AsyncSearchUSer into consideration ( which was missed when we introduced AsyncSearchUser - point in case )  and use that as a single source of truth for the check here ?

Reply:
Yes makes sense. Long logical expression is never good as an `if` condition and should be wrapped inside a method. 

But maybe we could find a better home for this method. How about [User](https://github.com/elastic/elasticsearch/blob/7.6/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/User.java)#isInternal(User user)?

The AsyncSearchUser is also missed in [InternalUserSerializationHelper](https://github.com/elastic/elasticsearch/blob/7.6/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/InternalUserSerializationHelper.java#L29). The new Use#isInternal method can be leveraged in here as well. It won't be able to access AuthorizationService#isInternalUser due to circular dependency.

Reply:
Sure thing!

Reply:
I did the refactoring. But decided to leave `InternalUserSerializationHelper` out for now. Changing it breaks backwards compatibility, which I think deseves its own issue.